packages/flutter/lib/src/rendering/object.dart

```diff
@@ -112,19 +112,27 @@ class PaintingContext extends ClipContext {
       );
       return true;
     }());
-    if (child._layer == null) {
+    OffsetLayer childLayer = child._layer;
+    if (childLayer == null) {
       assert(debugAlsoPaintedParent);
-      child._layer = OffsetLayer();
+      child._layer = childLayer = OffsetLayer();
```

Comment:
seems like this should go through the setter rather than the field, since we have a setter.

that said, it seems sketchy that we have two places that can affect this field. Who owns it, the render object or the painting context or someone else? we should have a clear ownership model so we know who's job it is to update it.

Reply:
Yeah, I couldn't think of a clean _and_ non-breaking solution here. We have two kinds of ownership, which is determined by `isRepaintBoundary`.

1. A repaint boundary does not own its layer. Instead, the `PaintingContext` creates an `OffsetLayer` and assigns it to the repaint boundary `RenderObject`.
1. A non-repaint boundary owns its layer (if it's composited, of course). In that case, the concrete implementation of the `RenderObject` would create the layer and assign it to the `layer` field.

To ensure correct usage, the public `layer` setter checks that the object is not a repaint boundary. Otherwise, one could create a repaint boundary and overwrite the layer assigned by `PaintingContext`. This is why I'm not using setter here.

Reply:
> I couldn't think of a clean and non-breaking solution here

In such cases the "non-breaking" part is the part we give up on. (See the breaking change policy.)

This sounds like a very confusing mess as implemented. I think we should fix this before landing. There should be a single clear documented owner for the field.

Reply:
(as discussed, it's fine if we just add a comment here that says we're setting the field directly because this is the one place that IS allowed to set it when it's a repaint boundary)

Reply:
Done